Python translation

emilybache opened the pull request:
The translation is based on the typescript version. The github action runs the tests.

nitsanavni commented on .github/workflows/python-build.yml:
```diff
@@ -0,0 +1,57 @@
+on:
+  push:
+    paths:
+      - 'python/**'
```
😎

nitsanavni commented on .github/workflows/python-build.yml:
```diff
@@ -0,0 +1,57 @@
+    - name: Checkout Repository
+      uses: actions/checkout@v2
+
+    - name: Start MYSQL and import DB
```
consider using https://github.com/marketplace/actions/actions-setup-mysql

emilybache replied:
This part is just copied from the typescript version. Should probably update all the versions together?

codecop replied:
I would leave that for a separate PR

nitsanavni replied:
I've found a cleaner way to do that:

https://github.com/nitsanavni/Refactoring-Kata-Lift-Pass-Pricing/pull/1/files#diff-88dc73ea041a254f6d251b3eb8872104b6926163e8a022e18a9b7d6939a99e73R23

nitsanavni commented on .github/workflows/python-build.yml:
```diff
@@ -0,0 +1,57 @@
+        sudo systemctl start mysql
+        mysqladmin --user=${{ env.DB_USER }} --password=${{ env.DB_OLD_PASSWORD }} version
+        mysqladmin --user=${{ env.DB_USER }} --password=${{ env.DB_OLD_PASSWORD }} password ${{ env.DB_PASSWORD }}
+        mysql -u${{ env.DB_USER }} -p${{ env.DB_PASSWORD }} < ${GITHUB_WORKSPACE}/database/initDatabase.sql  
```
like 👍🏼

nitsanavni commented on .github/workflows/python-build.yml:
```diff
@@ -0,0 +1,57 @@
+        mysqladmin --user=${{ env.DB_USER }} --password=${{ env.DB_OLD_PASSWORD }} password ${{ env.DB_PASSWORD }}
+        mysql -u${{ env.DB_USER }} -p${{ env.DB_PASSWORD }} < ${GITHUB_WORKSPACE}/database/initDatabase.sql  
+
+    - name: Install MySQL odbc driver
```
🤔 other translations don't rely on odbc, is it a must?

nitsanavni replied:
I've experimented with this alternative:
https://github.com/emilybache/Refactoring-Kata-Lift-Pass-Pricing/pull/1

emilybache replied:
I used odbc because it makes it easier to switch to a different database. It's probably not necessary here though as you point out

codecop commented:
Emily, thank you for the translation. Looks very good. There is a small inconsistency to the other projects - because also typescript version has it.

1. There is a `main` entry point (class, whatever) which calls `createApp` and then gets the connection and the application. It will print a message with a test URL and proper port (look into `index.ts` in the Typescript) and then starts the app by listening. So please instead of having the run of app in the main method of `Prices`, please create a separate file.

2. The `createApp` is also returning the connection, so tests can close it after running. I do not see the tests closing the connection, maybe a method provided to shutdown of process? I think it's OK to get the values `app` and importing by running the module like it is now, connection is just available globally then?

codecop commented:
3. I am missing the Readme that all other languages have. That should be little work, just copy any Readme and change links and commands to run the tests and application.

nitsanavni commented:
> I do not see the tests closing the connection

that's because the test is spawning the app as a child process, and once the child process is terminated it stops connections (at least, I believe it does)

martinsson commented:
Thanks @emilybache !

I tried to run this version on a macbook

The slight problem is I need to adapt the install of mysql-odbc. I think @nitsanavni suggestion solves that problem. Probably also for windows. However it's not complete (still some references to pyodbc). 

I'd suggest
- add the readme with the setup suggested by @codecop, I did the following (based on your github workflow)
```  
python3 -m venv venv
. venv/bin/activate
pip3 install -r requirements.txt
PYTHONPATH=src python -m pytest
```
- merge as-is noting that the setup is for linux

As a second step finalize @nitsanavni's contribution to make it platform independent.

emilybache commented on python/test/test_prices.py:
```diff
@@ -0,0 +1,79 @@
+        (15, 35, datetime.fromisoformat('2019-02-22')),
+        (15, 35, datetime.fromisoformat('2019-02-25')), # monday, holiday
+        (15, 23, datetime.fromisoformat('2019-03-11')), # monday
+        (65, 18, datetime.fromisoformat('2019-03-11')),  # monday
```
no I just translated the typescript tests.

emilybache commented:
Thankyou for all your comments! I feel a bit embarrassed I forgot the readme file, sorry about that. I'll fix that and look at your other comments soon.

codecop commented:
No I feel embarrassed to ask you for changes.

emilybache commented:
To answer the points from @codecop

1. I wasn't able to work out how to put the Flask application creation in a different file from the  methods with the app.route decoration. It might be possible but I don't know how. Sorry!
2. The db connection is not being explicitly closed, but it should close automatically when you kill the process. I don't think it's important to do explicitly. It is important that it's created in the same thread where it's used, and i fixed that at least.
3. I added a README which I hope does the job.

martinsson commented:
OK Great @emilybache ! this works on a macbook too using the normal dependency of docker. 

Nice try with sql lite as well although it fails at `ON DUPLICATE KEY UPDATE cost = ?`. I don't know how to solve this, other than going all the way to an ORM like sqlalchemy. This was actually the approach taken by https://github.com/martinsson/Refactoring-Kata-Lift-Pass-Pricing/pull/58/files. A MR I never got around to merge. I don't know what to do with it now though. It could be merged as a way of having sqllite. But If anyone has another idea of how to add the sqllite functionality to this project that'd be preferable.

Out of the suggestions you made [nitsanavni](https://github.com/nitsanavni). Is there any improvement you see still? Please to Provide a PR if so ! 👍

martinsson commented:
@emilybache FYI. I rewrote the test branch to alway keep the master branch mergeable into the with-tests branch. It's a convention that is useful when we make some modification to the master branche. 
It still looks like your commit as I authored it in your name. Ping me if you're not ok with that.

emilybache commented:
Hi,

Thanks Johan for fixing the git history so master can be merged into with_tests. I guess I was doing too much cherry-picking or something. 

As for the sqlite thing - I have this branch in my repo with a version of the production code that also works with sqlite: https://github.com/emilybache/Refactoring-Kata-Lift-Pass-Pricing/blob/with_tests_sqlite3/python/src/prices.py it's even more horrible than the original and of course doesn't match the other language versions. It is platform independent though.
